tools/wake-unit/fuzz_test_job_cache.cpp

```diff
@@ -293,6 +298,75 @@ struct Pool {
   }
 };
 
+template <class F>
+bool run_as_init_proc(F f) {
+  // We're going to unshare some namespaces but this can mess with later
+  // things that spawn threads so we want to fork first to isolate
+  int pid_wrapper = fork();
+
+  if (pid_wrapper == -1) {
+    wcl::log::error("run_as_init_proc: fork failure: %s", strerror(errno))();
+    return false;
+  }
+
+  // We wait for the child to return here
+  if (pid_wrapper != 0) {
+    int status = 0;
+    if (waitpid(pid_wrapper, &status, 0) != pid_wrapper) {
+      wcl::log::error("run_as_init_proc: waitpid(): %s", strerror(errno))();
+      return false;
+    }
+
+    // Relay errors up to the top
+    if (WIFEXITED(status) && WEXITSTATUS(status) == 0) {
+      return true;
+    }
+    return false;
+  }
+
+  // Now that we're in an isolated child, we unshare our namespaces
+  if (unshare(CLONE_NEWUSER | CLONE_NEWPID) != 0) {
+    wcl::log::error("unshare(CLONE_NEW_USER | CLONE_NEWPID): %s", strerror(errno))();
+    exit(1);
+  }
+
+  // Now the next child we create will be the init process for that namespace.
+  int pid = fork();
+
+  if (pid == -1) {
+    wcl::log::error("run_as_init_proc: fork failure: %s", strerror(errno))();
+    exit(1);
+  }
+
+  // In the child process we just want to run the function.
+  // We assure the user that the code executed by f() will be
+  // executed as the init process of its own namespace.
+  if (pid == 0) {
+    int retcode = f();
```

Comment:
A quirk of being PID1 is that other members of the pid namespace can not send you signals unless you choose to recieve them, so if you happen to be sending signals around it's something to be mindful of

Reply:
It should be fine, pid 1 shouldn't need to mess about with anything